[MRG+2] ENH: cross-validate over predefined folds

untom opened the pull request:
For some datasets, a pre-defined split of the data into several folds already exists. This is the case e.g. if one wants to build upon results produced by others, or even for famous datasets such as NORB. Currently, there is no way to use such predefined folds in sklearn when using e.g. GridSearchCV.

This PR adds `PredefinedSplit` as new CV-Split class that takes a predefined split and can be fed into GridSearchCV. 

Example usage:

```
# folds is an array that contains 0 for samples that are part of testfold 0, 1 for testfold 1, ....
cv = PredefinedSplit(folds)
gs = GridSearchCV(clf, param_grid=param_grid, cv=cv)
gs.fit(X, y)
```

Commit:
ENH: cross-validate over predefined splits

untom commented:
@jnothman can this get a MRG+2 ?

jnothman commented on sklearn/cross_validation.py:
```diff
@@ -1074,6 +1075,61 @@ def __len__(self):
+        test_fold[i] gives the test set fold of sample i. A value of -1
+        indicates that the corresponding sample is not part of any test set
+        folds, but will instead always be put into the training fold.
+
```
Please remove this blank line

jnothman commented on sklearn/cross_validation.py:
```diff
@@ -1074,6 +1075,61 @@ def __len__(self):
+    scheme. Each sample can be assigned to at most one test set fold, as
+    specified by the user through the ``test_fold`` parameter.
+
+
```
Please remove this blank line

jnothman commented:
Apart from the extra blank lines, this looks good to me.

untom commented:
done

untom commented:
I rebased on top of the current master to solve the merge commits, and squashed the commits. The Travis CI failure does not seem to be related to my changes (especially considering that the previous-to-last commit ran through and the last one just removed 2 empty lines).

amueller commented:
Meh, another random travis failure :-/

untom committed:
COSMIT pep8 cleanups

untom commented:
To make sure, I fixed a few unrelated pep8 warnings in the file and recommitted -- all green in CI now.

amueller commented:
Merging, then :)
